### PR TITLE
Try-state for Referenda pallet

### PR DESCRIPTION
**Pull request by Szegoo**
Part of: https://github.com/paritytech/polkadot-sdk/issues/239

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2698716

File: frame/referenda/src/lib.rs
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1328]: Any idea why is this happening?

File: frame/referenda/src/lib.rs
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1409]: Cargo fmt struggles with formatting this line.

**Comment by Szegoo** [comment]
@ggwpez Could you take a quick look?

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1298]: it can be, if it changed.

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1300]: same, it can be if changed

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1301]: if there is enough support and approves accumulated during preparation period, referendum will move into confirmation stage skipping the decision stage.

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1306]: duplication here

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1309]: I think this is possible when something is removed from the decision stage, and only with the next block the task from the queue will be removed.

File: frame/referenda/src/lib.rs
Comment (muharem) [inline frame/referenda/src/lib.rs:1311]:
```suggestion
	/// * The referendum indices stored in `TrackQueue` must exist as keys in the `ReferendumInfoFor`
```

File: frame/referenda/src/lib.rs
Comment (muharem) [reply frame/referenda/src/lib.rs:1328]: probably this does not return in order it was added, the StorageMap is a tree

File: frame/referenda/src/migration.rs
Comment (muharem) [inline frame/referenda/src/migration.rs:87]: this item not used, below you use the storage from the pallet.

Comment (Szegoo) [reply frame/referenda/src/migration.rs:87]: Yes, I initially used this but forgot to remove it later.

**Comment by Szegoo** [comment]
@muharem Thanks for your review, I am going to move this into draft now, I want to give this some more thought to see if something should be added/removed.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3040521

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3048548

File: frame/referenda/src/lib.rs
Comment (kianenigma) [inline frame/referenda/src/lib.rs:1306]: Why is this not a counted map?

File: frame/referenda/src/lib.rs
Comment (kianenigma) [reply frame/referenda/src/lib.rs:1320]: If in doubt, please remove it form the PR so this can be moved forward.

**Review by kianenigma — APPROVED** [review]
LGTM assuming that the existing assumptions all pass against Poladot and Kusama. A second pair of eyes by @gavofyork or @muharem would be great.

**Comment by Szegoo** [comment]
> LGTM assuming that the existing assumptions all pass against Poladot and Kusama.

I will run the checks against the Polkadot and Kusama storage and will share the results here.

File: frame/referenda/src/lib.rs
Comment (gpestana) [inline frame/referenda/src/lib.rs:1321]: fmt

File: frame/referenda/src/lib.rs
Comment (gpestana) [inline frame/referenda/src/lib.rs:1328]:
```suggestion
			"Number of referenda in `ReferendumInfoFor` is different than `ReferendumCount`"
```

**Review by gpestana — COMMENTED** [review]
I usually find that breaking down the logic in `try_state` separately is easier to read and understand what invariants are being checked. Especially because you often return a `DispatchResult` from each check closure within `do_try_runtime`. 

An example of breaking down the try runtime checks is implemented in the [EPM pallet](https://github.com/paritytech/substrate/blob/master/frame/election-provider-multi-phase/src/lib.rs#L1583). IMO this is not a deal breaker, but just thought of putting it out there for comparison and for you to decide which approach is better.

Thanks for helping here!

**Comment by Szegoo** [comment]
@gpestana I agree it does look nice when separated into multiple functions.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3081607

File: frame/referenda/src/lib.rs
Comment (Szegoo) [inline frame/referenda/src/lib.rs:1358]:
```suggestion
```
This is true for the first alarm, but the alarm can be set after the first nudge call(e.g. at the beginning of the deciding period). For this reason, this check has been failing against the Polkadot state and I am removing it.

**Comment by Szegoo** [comment]
Screenshot 2023-06-30 17_30_47

CC @kianenigma

**Comment by liamaharon** [comment]
@Szegoo could you also please check westend and rococo?

**Comment by Szegoo** [comment]
@liamaharon AFAIU they don't implement the referenda pallet.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285439

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285440

**Comment by liamaharon** [comment]
can you please rebase?

File: frame/referenda/src/migration.rs
Comment (liamaharon) [inline frame/referenda/src/migration.rs:197]: I think you can just do

```suggestion
	fn increment_referendum_index() {
        ReferendumCount::<T, ()>::mutate(|x| x.saturating_inc());
    }
```

so probably don't need `increment_referendum_index` and can do it inline

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285644

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3285629

File: frame/referenda/src/mock.rs
Comment (liamaharon) [inline frame/referenda/src/mock.rs]: formatting looks off here?

File: frame/referenda/src/lib.rs
Comment (liamaharon) [inline frame/referenda/src/lib.rs:1305]:
```suggestion
	/// * Referendum indices in [`MetadataOf`] must also be stored in [`ReferendumInfoFor`].
```

Comment (liamaharon) [reply frame/referenda/src/lib.rs:1305]: and in other comments please

File: frame/referenda/src/lib.rs
Comment (gpestana) [inline frame/referenda/src/lib.rs:1315]: Should we also check whether `ReferendumInfoFor` has metadata keys lingering? A count equality check would suffice after this look (ie. `ReferendumInfoFor::count == MetadataOf:::count`)

**Comment by gpestana** [comment]
bot fmt

**Review by gpestana — APPROVED** [review]
Left a comment but lgtm, thanks for the PR!

**Review by liamaharon — APPROVED** [review]

File: frame/referenda/src/lib.rs
Comment (Szegoo) [reply frame/referenda/src/lib.rs:1315]: `ReferendumInfoFor::count` won't always equal to `MetadataOf:::count`. The metadata is only set when calling the `set_metadata` extrinsic

**Comment by liamaharon** [comment]
bot merge

**Comment by liamaharon** [comment]
@paritytech/tip-bot-approvers could you please tip?

**Comment by juangirini** [comment]
@Szegoo could you please add your dot/ksm address to the PR description for a tip
format "[Polkadot|Kusama] address: xxx"

**Comment by juangirini** [comment]
@rzadp that's awesome, thanks for the update

**Pull request merged**
